cmd/soroban-rpc/internal/methods/get_ledgers.go

```diff
@@ -0,0 +1,251 @@
+	start := request.StartLedger
+	limit := h.defaultLimit
+	if request.Pagination == nil {
+		return start, limit, nil
+	}
+
+	var err error
+	if request.Pagination.Cursor != "" {
+		start, err = h.parseCursor(request.Pagination.Cursor, ledgerRange)
+		if err != nil {
+			return 0, 0, err
+		}
+	}
+	limit = h.handleLimit(request.Pagination.Limit)
+	return start, limit, nil
+}
+
+func (h ledgersHandler) parseCursor(cursor string, ledgerRange ledgerbucketwindow.LedgerRange) (uint32, error) {
+	cursorInt, err := strconv.ParseUint(cursor, 10, 32)
+	if err != nil {
+		return 0, err
+	}
+
+	start := uint32(cursorInt) + 1
+	if !isStartLedgerWithinBounds(start, ledgerRange) {
+		return 0, fmt.Errorf(
+			"cursor must be between the oldest ledger: %d and the latest ledger: %d for this rpc instance",
+			ledgerRange.FirstLedger.Sequence,
+			ledgerRange.LastLedger.Sequence,
+		)
+	}
+
+	return start, nil
+}
+
+func (h ledgersHandler) handleLimit(limit uint) uint {
+	if limit > 0 {
+		return limit
+	}
+	return h.defaultLimit
+}
+
+// fetchLedgers fetches ledgers from the DB for the range [start, start+limit-1]
+func (h ledgersHandler) fetchLedgers(ctx context.Context, start uint32,
+	limit uint, format string,
+) ([]LedgerInfo, error) {
+	ledgers, err := h.ledgerReader.BatchGetLedgers(ctx, start, limit)
+	if err != nil {
+		return nil, &jrpc2.Error{
+			Code:    jrpc2.InternalError,
+			Message: fmt.Sprintf("error fetching ledgers from db: %v", err),
+		}
+	}
+
+	// No more ledgers available
+	if len(ledgers) == 0 {
+		return nil, &jrpc2.Error{
+			Code:    jrpc2.InvalidParams,
+			Message: fmt.Sprintf("no ledgers found starting from sequence: %d", start),
+		}
+	}
+
+	result := make([]LedgerInfo, 0, limit)
+	for _, ledger := range ledgers {
+		if uint(len(result)) >= limit {
+			break
+		}
+
+		ledgerInfo, err := h.parseLedgerInfo(ledger, format)
+		if err != nil {
+			return nil, &jrpc2.Error{
+				Code:    jrpc2.InternalError,
+				Message: fmt.Sprintf("error processing ledger %d: %v", ledger.LedgerSequence(), err),
+			}
+		}
+		result = append(result, ledgerInfo)
+	}
+
+	return result, nil
+}
+
+// parseLedgerInfo extracts and formats the ledger metadata and header information.
+func (h ledgersHandler) parseLedgerInfo(ledger xdr.LedgerCloseMeta, format string) (LedgerInfo, error) {
+	ledgerInfo := LedgerInfo{
+		Hash:            ledger.LedgerHash().HexString(),
+		Sequence:        ledger.LedgerSequence(),
+		LedgerCloseTime: ledger.LedgerCloseTime(),
+	}
+
+	closeMetaB, err := ledger.MarshalBinary()
+	if err != nil {
+		return LedgerInfo{}, fmt.Errorf("error marshaling ledger close meta: %w", err)
+	}
+
+	headerB, err := ledger.LedgerHeaderHistoryEntry().MarshalBinary()
+	if err != nil {
+		return LedgerInfo{}, fmt.Errorf("error marshaling ledger header: %w", err)
+	}
+
+	// Format the data according to the requested format (JSON or XDR)
+	switch format {
+	case FormatJSON:
+		closeMetaJSON, headerJSON, convErr := ledgerToJSON(closeMetaB, headerB)
```

Comment:
You can actually get away with not marshaling ahead of time in the JSON case by using `xdr2json.ConvertInterface`, instead and just passing `ledger` and `ledger.LedgerHeaderHistoryEntry()` directly.

Reply:
Oooh thats nice! Thx

Reply:
@Shaptic On another look, I still need to marshal them to encode them for non-JSON response.
```
ledgerInfo.LedgerCloseMeta = base64.StdEncoding.EncodeToString(closeMetaB)
ledgerInfo.LedgerHeader = base64.StdEncoding.EncodeToString(headerB)
```
So I can just keep it